incubator/exporter-prometheus/src/main/java/io/aklivity/zilla/runtime/exporter/prometheus/internal/descriptor/PrometheusMetricDescriptor.java

```diff
@@ -0,0 +1,114 @@
+
+import static io.aklivity.zilla.runtime.engine.metrics.Metric.Kind.COUNTER;
+import static io.aklivity.zilla.runtime.engine.metrics.Metric.Unit.BYTES;
+
+import java.util.Map;
+
+import org.agrona.collections.Object2ObjectHashMap;
+
+import io.aklivity.zilla.runtime.engine.EngineConfiguration;
+import io.aklivity.zilla.runtime.engine.metrics.Metric;
+import io.aklivity.zilla.runtime.engine.metrics.MetricGroup;
+import io.aklivity.zilla.runtime.engine.metrics.MetricGroupFactory;
+
+public class PrometheusMetricDescriptor implements MetricDescriptor
+{
+    private final MetricGroupFactory metricGroupFactory;
+    private final Map<String, MetricGroup> metricGroups;
+    private final Map<String, String> names;
+    private final Map<String, String> kinds;
+    private final Map<String, String> descriptions;
+
+    public PrometheusMetricDescriptor(
+        EngineConfiguration config)
+    {
+        this.metricGroupFactory = MetricGroupFactory.instantiate();
+        this.metricGroups = metricGroups(config);
+        this.names = new Object2ObjectHashMap<>();
+        this.kinds = new Object2ObjectHashMap<>();
+        this.descriptions = new Object2ObjectHashMap<>();
+    }
+
+    @Override
+    public String kind(
+        String internalName)
+    {
+        String result = kinds.get(internalName);
+        if (result == null)
+        {
+            result = resolveMetric(internalName).kind().toString().toLowerCase();
+            kinds.put(internalName, result);
+        }
+        return result;
+    }
+
+    @Override
+    public String name(
+        String internalName)
+    {
+        String result = names.get(internalName);
+        if (result == null)
+        {
+            Metric metric = resolveMetric(internalName);
+            result = metric.name();
+            result = result.replace('.', '_');
+            if (metric.unit() == BYTES)
+            {
+                result += "_bytes";
+            }
+            if (metric.kind() == COUNTER)
+            {
+                result += "_total";
+            }
+            names.put(internalName, result);
+        }
+        return result;
+    }
+
+    @Override
+    public String description(
+        String internalName)
+    {
+        String result = descriptions.get(internalName);
+        if (result == null)
+        {
+            result = resolveMetric(internalName).description();
+            descriptions.put(internalName, result);
+        }
+        return result;
+    }
+
+    private Map<String, MetricGroup> metricGroups(
+        EngineConfiguration config)
+    {
+        final Map<String, MetricGroup> metricGroups = new Object2ObjectHashMap<>();
+        for (String name : metricGroupFactory.names())
+        {
+            MetricGroup metricGroup = metricGroupFactory.create(name, config);
+            metricGroups.put(name, metricGroup);
+        }
+        return metricGroups;
+    }
+
+    private Metric resolveMetric(
```

Comment:
Suggest moving `resolveMetric` to `EngineContext` as we'll need it for every `Exporter` implementation and it saves creating a new instance of all the metric group factories.

Reply:
done